program/rust/src/governance.rs

```diff
@@ -0,0 +1,198 @@
+
+    /// All these accounts are passed down and checked by the BPF loader
+    pub upgrade_authority: Derive<Info<'b>, UPGRADE_AUTHORITY_SEED>,
+    pub new_buffer:        Mut<Info<'b>>,
+    pub old_buffer:        Mut<Info<'b>>,
+    pub own_address:       Mut<Info<'b>>,
+    pub rent:              Sysvar<'b, Rent>,
+    pub clock:             Sysvar<'b, Clock>,
+    pub bpf_loader:        Info<'b>,
+    pub system:            Info<'b>,
+}
+
+
+pub fn upg_program(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    let cmd_args = load::<CommandHeader>(instruction_data)?;
+
+    // Solitaire checks
+    let mut accs: Box<UpgradeContract> = FromAccounts::from(program_id, &mut accounts.iter(), &())?;
+    // Solitaire context
+    let context: ExecutionContext = ExecutionContext {
+        program_id,
+        accounts,
+    };
+
+
+    // Pyth checks
+    check_valid_funding_account(&accs.funding_account)?;
+    check_valid_permissions_account(program_id, &accs.permissions_account)?;
+    let permissions_account_data =
+        load_checked::<PermissionAccount>(&accs.permissions_account, cmd_args.version)?;
+
+    // Verify source of governance
+    verify_governance(&accs.vaa, &permissions_account_data)?;
+    // Claim so no replayability
+    claim::consume(
+        &context,
+        accs.funding_account.key,
+        &mut accs.claim,
+        &accs.vaa,
+    )?;
+
+    let (upgrade_authority_pda, bump_seed) =
+        Pubkey::find_program_address(&[UPGRADE_AUTHORITY_SEED.as_bytes()], program_id);
+
+    let upgrade_ix = solana_program::bpf_loader_upgradeable::upgrade(
+        program_id,
+        &accs.vaa.new_contract,
+        &upgrade_authority_pda,
+        accs.funding_account.key,
+    );
+
+    invoke_signed(
+        &upgrade_ix,
+        accounts,
+        &[&[UPGRADE_AUTHORITY_SEED.as_bytes(), &[bump_seed]]],
+    )?;
+
+    Ok(())
+}
+
+
+/// Fail if the emitter is not the known governance key, or the emitting chain is not Solana.
+fn verify_governance<T>(
+    vaa: &PayloadMessage<T>,
+    permissions_account_data: &PermissionAccount,
+) -> ProgramResult
+where
+    T: DeserializePayload,
+{
+    let expected_emitter = permissions_account_data.master_authority;
+    let current_emitter = Pubkey::new_from_array(vaa.meta().emitter_address);
+    if expected_emitter != current_emitter || vaa.meta().emitter_chain != CHAIN_ID_SOLANA {
+        Err(OracleError::InvalidGovernanceVaa.into())
+    } else {
+        Ok(())
+    }
+}
+
+/// Payload for program upgrade
+pub struct PythGovernancePayloadUpgrade {
+    // Address of the new Implementation
+    pub new_contract: Pubkey,
+}
+
+impl SerializeGovernancePayload for PythGovernancePayloadUpgrade {
+    const MODULE: &'static str = "PythOracle";
+    const ACTION: u8 = 0;
+}
+
+impl SerializePayload for PythGovernancePayloadUpgrade {
+    fn serialize<W: Write>(&self, v: &mut W) -> std::result::Result<(), SolitaireError> {
```

Comment:
it seems that token bridge also writes the header in serialize method. [link](https://github.com/wormhole-foundation/wormhole/blob/dev.v2/solana/modules/token_bridge/program/src/messages.rs#L337)